src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -847,4 +847,138 @@ private void testNullHeadGetCigarBAM() {
+        Integer originalRefIndex = sam.getReferenceIndex();
+        Assert.assertTrue(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX != originalRefIndex);
+
+        sam.setHeader(null);
+        // force re-resolution of the reference name
+        sam.setHeaderStrict(samHeader);
+        Assert.assertEquals(sam.getReferenceIndex(), originalRefIndex);
+    }
+
+    @Test
+    public void testSetHeaderStrictValidNewHeader() {
+        final SAMRecord sam = createTestRecordHelper();
+        final String origSequenceName = sam.getContig();
+
+        final SAMFileHeader origSamHeader = sam.getHeader();
+        final int origSequenceLength = origSamHeader.getSequence(origSequenceName).getSequenceLength();
+        final SAMFileHeader newHeader = new SAMFileHeader();
+        newHeader.addSequence(new SAMSequenceRecord(origSequenceName, origSequenceLength));
+
+        // force re-resolution of the reference name against the new header
+        sam.setHeaderStrict(newHeader);
+        Assert.assertEquals(sam.getReferenceIndex(), new Integer(0));
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testSetHeaderStrictInvalidReference() {
+        SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
+
+        sam.setReferenceName("unresolveable");
+        Assert.assertEquals(new Integer(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX), sam.getReferenceIndex());
+
+        // throw on force re-resolution of the unresolveable reference name
+        sam.setHeaderStrict(samHeader);
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testSetHeaderStrictInvalidMateReference() {
+        SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
+
+        sam.setMateReferenceName("unresolveable");
+        Assert.assertEquals(new Integer(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX), sam.getMateReferenceIndex());
+
+        // throw on force re-resolution of the unresolveable mate reference name
+        sam.setHeaderStrict(samHeader);
+    }
+
+    @Test
+    public void testSetHeaderStrictNull() {
+        SAMRecord sam = createTestRecordHelper();
+        sam.setHeaderStrict(null);
+        Assert.assertEquals(sam.mReferenceIndex, null);
+    }
+
+    // resolveIndexFromName
+
+    @Test
+    public void testResolveIndexResolveable() {
+        final SAMRecord sam = createTestRecordHelper();
+        final SAMFileHeader samHeader = sam.getHeader();
+        final String contigName = sam.getContig();
+        Assert.assertEquals(SAMRecord.resolveIndexFromName(contigName, samHeader, true), new Integer(samHeader.getSequenceIndex(contigName)));
+    }
+
+    @Test(expectedExceptions=IllegalStateException.class)
+    public void testResolveIndexUnresolveableStrict() {
+        Assert.assertEquals(SAMRecord.resolveIndexFromName("unresolveable", null, false), new Integer(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
```

Comment:
Test case is mis-named (strict mode is off here). This is testing whether you throw when the header is null and strict mode is off. You also need a separate test with an unresolvable contig, a non-null header, and strict mode on that expects an exception.